### ISSUE TITLE
No autocompletions for local variables

### ISSUE THREAD

**Issue by mansimarkaur** [jpsim/SourceKitten#406]
```sourcekitten complete --text "let hello = 5;hell" --offset 18``` returns `[]`
@jpsim @norio-nomura  Is there a workaround for this or am I doing something wrong?

**Comment by jpsim** [comment]
tl;dr; You need to pass the byte offset to the first character of the string you want completed.

```bash
$ sourcekitten complete --text "let hello = 5;hell" --offset 14
```

---

In my opinion, SourceKit (the Apple-built library that implements most of the functionality exposed via SourceKitten) has very unintuitive & unforgiving code completion behavior.

You'll find examples of valid & invalid code completion requests in Swift's tests: https://github.com/apple/swift/tree/master/test/SourceKit/CodeComplete

You'll also find working examples in SourceKitten itself:

* In the "Complete" section of the README: https://github.com/jpsim/SourceKitten#complete
* In the "CodeCompleteion" unit tests: https://github.com/jpsim/SourceKitten/blob/master/Tests/SourceKittenFrameworkTests/CodeCompletionTests.swift

---

If I modify your example a bit by changing the offset, I can get the `hello` code completion item to be returned (using [`jq`](https://stedolan.github.io/jq/) to filter the json):

```bash
$ sourcekitten complete --text "let hello = 5;hell" --offset 14 | jq '.[] | select(.name | contains("hello"))'
{
  "descriptionKey": "hello",
  "kind": "source.lang.swift.decl.var.global",
  "name": "hello",
  "sourcetext": "hello",
  "numBytesToErase": 0,
  "typeName": "Int",
  "associatedUSRs": "s:v4main5helloSi",
  "moduleName": "main",
  "context": "source.codecompletion.context.local"
}
```

**Comment by mansimarkaur** [comment]
`sourcekitten complete --text "let hello = 5;hell" --offset 14 | jq ".[].name" | wc -l` returns 344
And out of these 344 results, there's only one which even starts with h and that is `hello`.
Why are wrong autocompletions being generated alongwith only one correct one?

**Comment by jpsim** [comment]
I don't know, that'd be a great question for the swift-dev mailing list! https://lists.swift.org/mailman/listinfo/swift-dev

Let me know what they say! 😄

**Comment by ainopara** [comment]
I think that is reasonable since the work of filter / sort code complete results is belongs to IDE.

By assigning offset to 14:
1. SourceKit ignore texts you already typed (in this case `hell`) and return all 
possible value in this context. 
2. Then IDE will filter it basing on text between offset 14 and current cursor offset.  

By pressing `ESC` you can see that Xcode will give code completion suggest on code that already completed which fits this workflow.
![image](https://user-images.githubusercontent.com/1849450/31050889-d3a4f9f8-a61e-11e7-936c-4d0cce02a771.png)

**Comment by jpsim** [comment]
@mansimarkaur did you ever start a conversation on swift-dev about this? Did you ever hear back? Closing since this issue hasn't seen any activity in a while.

Thanks for sharing your thoughts @ainopara.

**Comment by ainopara** [comment]
It seems @mansimarkaur did post a question on swift-dev. https://forums.swift.org/t/autocompletions-for-local-variables/6374
I happened to find that by searching `SourceKit` in the new Discourse based forum. That's pretty useful. 😋
